src/Microsoft.AspNetCore.TestHost/HttpContextBuilder.cs

```diff
@@ -0,0 +1,125 @@
+            var requestLifetimeFeature = new HttpRequestLifetimeFeature();
+            requestLifetimeFeature.RequestAborted = _requestAbortedSource.Token;
+            _httpContext.Features.Set<IHttpRequestLifetimeFeature>(requestLifetimeFeature);
+            
+            _responseStream = new ResponseStream(ReturnResponseMessageAsync, AbortRequest);
+            _responseFeature.Body = _responseStream;
+        }
+
+        internal void Configure(Action<HttpContext> configureContext)
+        {
+            if (configureContext == null)
+            {
+                throw new ArgumentNullException(nameof(configureContext));
+            }
+
+            configureContext(_httpContext);
+        }
+
+        /// <summary>
+        /// Start processing the request.
+        /// </summary>
+        /// <returns></returns>
+        internal async Task<HttpContext> SendAsync(CancellationToken cancellationToken = default)
+        {
+            var registration = cancellationToken.Register(AbortRequest);
+
+            _testContext = _application.CreateContext(_httpContext.Features);
+
+            // Async offload, don't let the test code block the caller.
+            var offload = Task.Factory.StartNew(async () =>
+            {
+                try
+                {
+                    await _application.ProcessRequestAsync(_testContext);
+                    await CompleteResponseAsync();
+                    _application.DisposeContext(_testContext, exception: null);
+                }
+                catch (Exception ex)
+                {
+                    Abort(ex);
+                    _application.DisposeContext(_testContext, ex);
+                }
+                finally
+                {
+                    registration.Dispose();
+                }
+            });
+
+            return await _responseTcs.Task.ConfigureAwait(false);
+        }
+
+        internal void AbortRequest()
+        {
+            if (!_pipelineFinished)
+            {
+                _requestAbortedSource.Cancel();
+            }
+            _responseStream.Complete();
+        }
+
+        internal async Task CompleteResponseAsync()
+        {
+            _pipelineFinished = true;
+            await ReturnResponseMessageAsync();
+            _responseStream.Complete();
+            await _responseFeature.FireOnResponseCompletedAsync();
+        }
+
+        internal async Task ReturnResponseMessageAsync()
+        {
+            // Check if the response has already started because the TrySetResult below could happen a bit late
+            // (as it happens on a different thread) by which point the CompleteResponseAsync could run and calls this
+            // method again.
+            if (!_responseFeature.HasStarted)
+            {
+                // Sets HasStarted
+                await _responseFeature.FireOnSendingHeadersAsync();
+                var setResult = _responseTcs.TrySetResult(_httpContext);
```

Comment:
what is setResult used for?

Reply:
To prevent the compiler warning about ignoring return values

Reply:
There's a first class feature called "discards" for that now.

```C#
_ = _responseTcs.TrySetResult(_httpContext);
```